### ISSUE TITLE
Support loose event matching in Test Kit

### ISSUE THREAD

**Issue by jlink** [junit-team/junit5#1771]
Follow up of #1532

## Description

Provide method to assert events regardless of completeness and order.

## Example

```java
Events events = 
    EngineTestKit
	.engine(testEngine)
	.execute()
	.all();

    events.assertEventsInAnyOrder(
	event(container(MyContainerClass.class), started()),
	...
	event(container(MyContainerClass.class), finishedSuccessfully())
    );
```

### Why should that be part of the Test Kit?

My guess is that most test engine builders have the need to check for only some of the generated events. My attempts to use `events.assertEventsMatch()` did not lead anywhere since AssertJ's `ListAssert` class does not come with something like `matchOneOf(Condition... conditions)`.

**Comment by sbrannen** [comment]
> My guess is that most test engine builders have the need to check for only some of the generated events.

Although perhaps not the most intuitive (due to the AssertJ APIs), this is already possible via the `Events#assertThatEvents()` method.

> My attempts to use `events.assertEventsMatch()` did not lead anywhere since AssertJ's `ListAssert` class does not come with something like `matchOneOf(Condition... conditions)`.

Yes, I believe that's correct that AssertJ does not support a single such method, but as an _interim_ solution you can chain together multiple of the `have*()` variants (from `AbstractIterableAssert`) to achieve that.

- `haveAtLeastOne()`
- `haveAtLeast()`
- `haveAtMost()`
- `haveExactly()`

For example, the following snippet taken from `ParameterizedTestIntegrationTests` in the JUnit 5 test suite achieves that.

```java
var results = execute(selectClass(LifecycleTestCase.class));
results.all().assertThatEvents()
        .haveExactly(1, event(test("test1"), displayName("[1] foo"), finishedWithFailure(message("foo"))))
        .haveExactly(1, event(test("test1"), displayName("[2] bar"), finishedWithFailure(message("bar"))));
```

**Comment by sbrannen** [comment]
I agree that having to chain multiple invocations of `haveExactly()` and always having to supply `1` as the first argument is cumbersome.

I therefore propose we introduce the following method in `Events` with the semantics proposed by @jlink.

```java
@SafeVarargs
public final void assertEventsMatch(Condition<? super Event>... conditions) {
	Preconditions.notNull(conditions, "conditions must not be null");
	// TODO
}
```

**Comment by sbrannen** [comment]
Tentatively slated for inclusion in 5.5 M1

**Comment by jlink** [comment]
Any news on that one?

**Comment by marcphilipp** [comment]
**Team decision:** Add `assertEventsMatchLoosely` and document that it does a partial match in any order.

**Comment by mmerdes** [comment]
in progress

**Comment by marcphilipp** [comment]
Resolved by #2250.
